compiler/rustc_trait_selection/src/traits/wf.rs

```diff
@@ -421,13 +414,44 @@ impl<'tcx> WfPredicates<'tcx> {
         let obligations = self.nominal_obligations_without_const(data.def_id, data.substs);
         self.out.extend(obligations);
 
+        self.compute_projection_substs(data.substs);
+    }
+
+    fn compute_inherent_projection(&mut self, data: ty::AliasTy<'tcx>) {
+        // An inherent projection is well-formed if
+        //
+        // (a) its predicates hold (*)
+        // (b) its substs are wf
+        //
+        // (*) The predicates of an inherent associated type include the
+        //     predicates of the impl that it's contained in.
+
+        if !data.self_ty().has_escaping_bound_vars() {
+            // FIXME(inherent_associated_types): Should this happen inside of a snapshot?
+            let substs = traits::project::compute_inherent_assoc_ty_substs(
```

Comment:
Hmmmm... this is the only thing that give me a bit of anxiety. Perhaps we shouldn't normalize the associated type here, and simply just compute it and its nominal obligations here. :thinking:

Reply:
Yeah, and this will be needed to do WF properly once we have lazy norm too. Since we're not normalizing, we'd also need to register the GAT predicates here as WF obligations, I think.

Reply:
I'm also wary about adding this much complexity here. Conceptually I "only" want to obtain the "normal" substs `[I0..In, P0..Pm]` from the IAT substs `[Self, P0..Pm]` to be able to call `nominal_obligations()` which calls `tcx.predicates_of` (which obv requires parent substs, here `I0..Im`). In reality that's not that simple.

I already account for the GAT substs. They are checked for wfness below: `compute_projection_substs`.

Reply:
Re. normalizing, `compute_*_substs()` doesn't normalize the inherent projection but it does normalize the impl ty (with depth) to mirror IAT selection. Otherwise they wouldn't agree. I can not do that but not sure what's gonna happen (and in such case we defo need to run inside of a snapshot since we then try to unify the self ty with the unnorm impl ty, hmmm)

Reply:
Oh, I see now. Yeah, you need to normalize the self type of the impl as well. Gah, that's frustrating :facepalm:

Reply:
I guess you can just leave a FIXME here that it's not going to be compatible with the new trait solver and lazy norm? This shouldn't be forgotten :smile:

Reply:
Can do. And you are saying that I should keep this latest commit, right?